cni/pkg/install-cni/pkg/util/util.go

```diff
@@ -0,0 +1,170 @@
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+)
+
+func CopyAtomically(srcFilepath, targetDir, targetFilename string) error {
+	info, err := os.Stat(srcFilepath)
+	if err != nil {
+		return err
+	}
+
+	input, err := ioutil.ReadFile(srcFilepath)
+	if err != nil {
+		return err
+	}
+
+	return WriteAtomically(filepath.Join(targetDir, targetFilename), input, info.Mode())
+}
+
+// Write atomically by writing to a temporary file in the same directory then renaming
+func WriteAtomically(path string, data []byte, mode os.FileMode) (err error) {
+	tmpFile, err := ioutil.TempFile(filepath.Dir(path), filepath.Base(path)+".tmp.")
+	if err != nil {
+		return
+	}
+	defer func() {
+		if fileutil.Exist(tmpFile.Name()) {
+			if rmErr := os.Remove(tmpFile.Name()); rmErr != nil {
+				if err != nil {
+					err = errors.Wrap(err, rmErr.Error())
+				} else {
+					err = rmErr
+				}
+			}
+		}
+	}()
+
+	if err = os.Chmod(tmpFile.Name(), mode); err != nil {
+		return
+	}
+
+	_, err = tmpFile.Write(data)
+	if err != nil {
+		if closeErr := tmpFile.Close(); closeErr != nil {
+			err = errors.Wrap(err, closeErr.Error())
+		}
+		return
+	}
+	if err = tmpFile.Close(); err != nil {
+		return
+	}
+
+	err = os.Rename(tmpFile.Name(), path)
+	return
+}
+
+func CreateFileWatcher(dir string) (watcher *fsnotify.Watcher, fileModified chan bool, errChan chan error, err error) {
+	watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return
+	}
+
+	fileModified, errChan = make(chan bool), make(chan error)
+	go WatchFiles(watcher, fileModified, errChan)
+
+	if err = watcher.Add(dir); err != nil {
+		if closeErr := watcher.Close(); closeErr != nil {
+			err = errors.Wrap(err, closeErr.Error())
+		}
+		return nil, nil, nil, err
+	}
+
+	return
+}
+
+func WatchFiles(watcher *fsnotify.Watcher, fileModified chan bool, errChan chan error) {
+	for {
+		select {
+		case _, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+			fileModified <- true
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return
+			}
+			errChan <- err
+		}
+	}
+}
+
+func WaitForFileMod(ctx context.Context, fileModified chan bool, errChan chan error) error {
+	select {
+	case <-fileModified:
+		return nil
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func ReadCNIConfigMap(path string) (map[string]interface{}, error) {
+	cniConfig, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var cniConfigMap map[string]interface{}
+	if err = json.Unmarshal(cniConfig, &cniConfigMap); err != nil {
+		return nil, errors.Wrap(err, path)
+	}
+
+	return cniConfigMap, nil
+}
+
+func GetPlugins(cniConfigMap map[string]interface{}) (plugins []interface{}, err error) {
+	defer func() {
+		if panicErr := recover(); panicErr != nil {
+			plugins = nil
+			err = errors.New("error reading plugin list from CNI config")
+		}
+	}()
+	plugins = cniConfigMap["plugins"].([]interface{})
```

Comment:
`plugins, ok := cniConfigMap["plugins"].([]interface{})` to avoid recover from panic.

Reply:
I believe the panic could still occur with `.([]interface{})`

Reply:
If you specify `ok`, when type conversion fails, it will assign `ok=false`, instead of panicking.

Reply:
Ah I was not aware. I thought you were referring to accessing the map. Thanks for the explanation!